stdlib/LinearAlgebra/src/eigen.jl

```diff
@@ -275,6 +275,18 @@ function eigen(A::AbstractMatrix{T}; permute::Bool=true, scale::Bool=true, sortb
     isdiag(AA) && return eigen(Diagonal(AA); permute=permute, scale=scale, sortby=sortby)
     return eigen!(AA; permute=permute, scale=scale, sortby=sortby, jvl=jvl, jvr=jvr, jce=jce, jcv=jcv)
 end
+function eigen(A::AbstractMatrix{T}; permute::Bool=true, scale::Bool=true, sortby::Union{Function,Nothing}=eigsortby) where {T <: Union{Float16,Complex{Float16}}}
+    AA = copy_oftype(A, eigtype(T))
+    isdiag(AA) && return eigen(Diagonal(AA); permute=permute, scale=scale, sortby=sortby)
+    A = eigen!(AA; permute=permute, scale=scale, sortby=sortby)
+    if isreal(A.values)
+        Eigen(convert(Vector{Float16}, A.values),
+            convert(AbstractArray{Float16}, A.vectors))
+    else
+        Eigen(convert(Vector{Complex{Float16}}, A.values),
+            convert(AbstractArray{Complex{Float16}}, A.vectors))
+    end
```

Comment:
A general conversion constructor seems to be too involved. Here we are in a very specific context, so the following should hopefully do the job:
```suggestion
    values = convert(AbstractVector{isreal(A.values) ? Float16 : Complex{Float16}}, A.values)
    vectors = convert(AbstractMatrix{isreal(A.vectors) ? Float16 : Complex{Float16}}, A.vectors)
    vectorsl = convert(AbstractMatrix{isreal(A.vectorsl) ? Float16 : Complex{Float16}}, A.vectorsl)
    return Eigen(values, vectors, vectorsl, A.unitary, A.rconde, A.rcondv)
```

Reply:
I just added the type converter. If its good to implement, we can use that instead?

Reply:
Something like this would do the job I guess. 
```
function eigen(A::AbstractMatrix{T}; permute::Bool=true, scale::Bool=true, sortby::Union{Function,Nothing}=eigsortby) where {T <: Union{Float16,Complex{Float16}}}
    AA = copy_oftype(A, eigtype(T))
    isdiag(AA) && return eigen(Diagonal(AA); permute=permute, scale=scale, sortby=sortby)
    A = eigen!(AA; permute=permute, scale=scale, sortby=sortby)
    return Eigen{T}(A)
end
```

Reply:
Nvm, there are corner cases where it does not work. Like a real matrix with complex eigen values.